photon-ml/src/main/scala/com/linkedin/photon/ml/function/SmoothedHingeLossFunction.scala

```diff
@@ -14,19 +14,83 @@
  */
 package com.linkedin.photon.ml.function
 
-import breeze.linalg
-import com.linkedin.photon.ml.data.LabeledPoint
+import breeze.linalg.Vector
+import com.linkedin.photon.ml.data.{LabeledPoint, SimpleObjectProvider}
+import com.linkedin.photon.ml.normalization.{NoNormalization, NormalizationContext}
+import com.linkedin.photon.ml.util.Utils
+import org.apache.spark.rdd.RDD
 
 /**
- * Implement Rennie's smoothed hinge loss function (http://qwone.com/~jason/writing/smoothHinge.pdf) as an
- * optimizer-friendly approximation for linear SVMs
- */
-class SmoothedHingeLossFunction extends DiffFunction[LabeledPoint] {
+  * Implement Rennie's smoothed hinge loss function (http://qwone.com/~jason/writing/smoothHinge.pdf) as an
+  * optimizer-friendly approximation for linear SVMs
+  *
+  * FAQ: Why use cumGradient (cumulative gradient)?
+  * A: Using cumGradient allows the functions to avoid memory allocation by modifying and returning cumGradient instead
+  *    of creating a new gradient vector.
+  *
+  * @param data The given data at which to compute the hessian multiplied by a given vector
+  * @param treeAggregateDepth The depth used by treeAggregate. Depth 1 indicates normal linear aggregate. Using
+  *                           depth > 1 can reduce memory consumption in the Driver and may also speed up the
+  *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+  *                           versions 1.4 and 1.5.
+  */
+class SmoothedHingeLossFunction(
```

Comment:
@joshvfleming 
The default behaviour code that was in DiffFunction was only used by the SmoothedHingeLossFunction and overwritten by the GLMLossFunction, so I moved it into the SmoothedHingeLossFunction and as previously mentioned turned DiffFunction into a pure interface.